academy/ibc-dev/ibcapp_steps.md

```diff
@@ -0,0 +1,425 @@
+
+Modules may return to the IBC handler an acknowledgement which implements the `Acknowledgement` interface.
+The IBC handler will then commit this acknowledgement of the packet so that a relayer may relay the acknowledgement back to the sender module.
+
+The state changes that occurred during this callback will only be written if:
+
+- the acknowledgement was successful as indicated by the `Success()` function of the acknowledgement
+- if the acknowledgement returned is nil indicating that an asynchronous process is occurring
+
+**NOTE:** Applications which process asynchronous acknowledgements must handle reverting state changes when appropriate. Any state changes that occurred during the `OnRecvPacket` callback will be written for asynchronous acknowledgements.
+
+In `x/leaderboard/module_ibc.go` scaffolded by Ignite CLI we find `OnRecvPacket`:
+
+```go
+// OnRecvPacket implements the IBCModule interface
+func (am AppModule) OnRecvPacket(
+	ctx sdk.Context,
+	modulePacket channeltypes.Packet,
+	relayer sdk.AccAddress,
+) ibcexported.Acknowledgement {
+	var ack channeltypes.Acknowledgement
+
+	// this line is used by starport scaffolding # oracle/packet/module/recv
+
+	var modulePacketData types.LeaderboardPacketData
+	...
+
+	// Dispatch packet
+	switch packet := modulePacketData.Packet.(type) {
+	// this line is used by starport scaffolding # ibc/packet/module/recv
+	default:
+		errMsg := fmt.Sprintf("unrecognized %s packet type: %T", types.ModuleName, packet)
+		return channeltypes.NewErrorAcknowledgement(errMsg)
+	}
+
+	// NOTE: acknowledgement will be written synchronously during IBC handler execution.
+	return ack
+}
+```
+The *dispatch packet* switch statement is added by Ignite CLI, as it is stated in the docs we strictly speaking only need to decode the packet data (which we'll discuss in a next section) and return the acknowledgement after processing the packet. The structure provided by Ignite CLI is however useful to get us set up, but can be changed according to the preference of the developer.
+
+As a reminder, the `Acknowledgement` interface:
+
+```go
+// Acknowledgement defines the interface used to return
+// acknowledgements in the OnRecvPacket callback.
+type Acknowledgement interface {
+	Success() bool
+	Acknowledgement() []byte
+}
+```
+
+##### Acknowledging packets
+
+The last step of the packet flow depends on whether we have a happy path when the packet has been successfully relayed or a timeout when something went wrong.
+
+After a module writes an acknowledgement, a relayer can relay back the acknowledgement to the sender module. The sender module can
```

Comment:
spacing issue here?

Reply:
Not sure what you mean? The extra line after the first sentence? This I added because after the first line the reader may want to jump straight to timeouts so I felt it made sense to add this

I think we can leave this for the language review ;)

Reply:
oh I see what you mean now